Enhance diagnostics docs (#15648)
gravitational/teleport: docs/pages/management/diagnostics/tracing.mdx

---
title: Distributed Tracing
description: How to enable tracing within Teleport.
---

Teleport leverages [OpenTelemetry](https://opentelemetry.io/) to generate traces
and export them to any [OpenTelemetry Protocol (OTLP)](https://opentelemetry.io/docs/reference/specification/protocol/otlp/)
capable exporter. In the event that your telemetry backend doesn't support receiving OTLP traces, you may be able to
leverage the [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/) to proxy traces from OTLP
to a format that your telemetry backend accepts.

## Configure Teleport

> In order to enable tracing in Teleport, add the following section in `teleport.yaml`. For a detailed description of

Reviewer: ```suggestion
In order to enable tracing for a `teleport` instance, add the following 
section to that instance's configuration file (`/etc/teleport.yaml` by default). 
For a detailed description of
```

these configuration fields, see the [configuration reference](./reference/configuration.mdx) page.

> <Details title="Tracing Service Configuration" min="8.3.18,9.3.15,10.1">

Reviewer: I don't think a Details box is the right component here, since it's collapsed by default and the "Configure Teleport" section revolves around it. I would use regular text instead. 

If we want to warn users about minimum versions, I think we should be explicit about this in the body text so it's clear that there are limitations.

```yaml
tracing_service:
   enabled: yes
   exporter_url: grpc://collector.example.com:4317
   sampling_rate_per_million: 1000000
```
</Details>

It is important to choose the sampling rate wisely. Sampling at a rate of 100% could have a negative impact on the

> performance of your cluster. Teleport honors the parent span's sampling rate, which means that even when the

Reviewer: We should introduce "the parent span", since it's not as clear as it could be what this is referring to. Maybe, "When generating data for a span within a trace, Teleport honors...".

`tracing_service` is enabled and the sampling rate is 0, if Teleport receives a request that has a span which is

> sampled, then Teleport will sample and exports all spans that are generated in response to that request.

Reviewer: ```suggestion
sampled, then Teleport will sample and export all spans that are generated in response to that request.
```

> After updating `teleport.yaml`, start Teleport as usual using `teleport start`.

Reviewer: ```suggestion
After updating `teleport.yaml`, start your `teleport` instance to apply the new configuration.
```
Suggesting a vaguer phrasing since users who installed Teleport via package manager on systemd-enabled systems would be using systemctl.

---
title: Distributed Tracing
description: How to enable tracing within Teleport.
---

Teleport leverages [OpenTelemetry](https://opentelemetry.io/) to generate traces
and export them to any [OpenTelemetry Protocol (OTLP)](https://opentelemetry.io/docs/reference/specification/protocol/otlp/)

> capable exporter. In the event that your telemetry backend doesn't have support receiving OTLP traces you may be able to

Reviewer: ```suggestion
capable exporter. In the event that your telemetry backend doesn't support receiving OTLP traces, you may be able to
```

leverage the [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/) to proxy traces from OTLP
to a format that your telemetry backend accepts.

## Configure Teleport

> In order to enable tracing in Teleport, add the following section in `teleport.yaml`. For a detailed description of

Reviewer: Is this the `teleport.yaml` for a specific `teleport` instance you want to enable tracing for, or do you only need to enable tracing on one `teleport` instance, e.g., the Auth or Proxy Service?

these configuration fields, see the [configuration reference](./reference/configuration.mdx) page.

> <Details title="Tracing Service Configuration" min="8.3.18,9.3.15,10.1" scopeOnly={true} scope={["oss", "enterprise"]}>

Reviewer: Readers with the `cloud` will see the first paragraph of this section, then the Admonition. 

Is this only something you can set on the Auth/Proxy service? In that case, we should explain that in a Prerequisites section.

We should also remove this code fence from the `Details` box, since it should be equally visible for all readers of the guide, and I'm not sure there's a reason to collapse it.

```yaml
tracing_service:
   enabled: yes

>    exporter_url: grpc://collector.example.com:4317

Reviewer: Is there anything we should know about how Teleport parses this URL? E.g.,

- Which schemes are supported?
- Is a scheme required?

Reviewer: I think so, since we can't count on users clicking the link above.

>    sampling_rate_per_million: 1000000

Reviewer: What's the default for this? Is it reasonable to leave it unset?

```
</Details>

> <Admonition type="warning" title="Sampling Rate">

Reviewer: Since there's so little text below the configuration snippet, I think we can turn this Admonition into regular body text.

> It is important to chose the sampling rate wisely. Sampling at a rate of 100% could have a negative impact on the

Reviewer: ```suggestion
It is important to choose the sampling rate wisely. Sampling at a rate of 100% could have a negative impact on the
```

> performance of your cluster. Teleport honors the parent spans sampling rate, which means even when the the

Reviewer: ```suggestion
performance of your cluster. Teleport honors the parent span's sampling rate, which means that even when the
```

> performance of your cluster. Teleport honors the parent spans sampling rate, which means even when the the

Reviewer: This is the first mention of "parent span", which leads me to wonder how Teleport structures traces. Is each span a separate gRPC message, or is there some other way traces are structured?

> `tracing_service` is enabled and the sampling rate is 0 if Teleport receives a request that has a span which is

Reviewer: ```suggestion
`tracing_service` is enabled and the sampling rate is 0, if Teleport receives a request that has a span which is
```

sampled, then Teleport will sample and exports all spans that are generated in response to that request.
</Admonition>

> After updating `teleport.yaml`, start Teleport as usual using `teleport start`.

Reviewer: Add a systemctl command here?

## tsh

To capture traces from `tsh` simply add the `--trace` flag to your command. All traces generated by `tsh --trace` will be

> proxied to the `exporter_url` defined for the Auth service of the cluster the command is being run on.

Reviewer: ```suggestion
proxied to the `exporter_url` defined for the Auth Service of the cluster the command is being run on.
```
